Sources/EpoxyCollectionView/CollectionView/ReusableViews/CollectionViewCell.swift

```diff
@@ -120,22 +120,31 @@ public final class CollectionViewCell: UICollectionViewCell, ItemCellView {
     // On iOS 15, cells that return an unstable size can result in a layout recursion crash. The
     // root cause of these crashes is likely an ambiguous layout due to misconfigured constraints.
     // Unfortunately, finding each and every one of these ambiguous layouts can be challenging. This
-    // code works around this new behavior / crash by giving up on self-sizing after 5 attempts.
-    // Once a component has been sized differently 5 times, we assume that it has an ambiguous
-    // layout and we simply return the last computed size for the remainder of the component's life.
-    // We reset the count on cell reuse and when the cell is reconfigured in `CollectionView`'s
-    // `configure` function.
-    if CollectionViewConfiguration.shared.enableLayoutRecursionWorkaround {
-      if let previousComputedSize = previousComputedSize, numberOfNewComputedSizes >= 5 {
-        EpoxyLogger.shared.assertionFailure(
-          "Layout recursion detected. View: \(view?.description ?? "nil view"). Size: \(preferredAttributes.size).")
-        preferredAttributes.size = previousComputedSize
+    // code tries to detect these layout issues and logs a warning when one is detected: if a
+    // component size is unstable for at least 50 sizing attempts in less than 2 seconds, we log a
+    // warning.
+
+    let sizingAttemptTime = CACurrentMediaTime()
+    let wasFirstSizingAttemptRecent: Bool
+    if let firstSizingAttemptTime = firstSizingAttemptTime {
+      wasFirstSizingAttemptRecent = sizingAttemptTime - firstSizingAttemptTime < 2
+      if !wasFirstSizingAttemptRecent {
+        // If the last sizing attempt was not nil, but also more than the 2s threshold, reset all
+        // self-sizing-layout-recursion-tracking state.
+        resetSelfSizingLayoutRecursionTrackingState()
       }
+    } else {
+      firstSizingAttemptTime = sizingAttemptTime
+      wasFirstSizingAttemptRecent = false
+    }
 
-      if preferredAttributes.size != previousComputedSize {
-        numberOfNewComputedSizes += 1
-        previousComputedSize = preferredAttributes.size
-      }
+    if wasFirstSizingAttemptRecent, previousComputedSizes.count >= 50 {
```

Comment:
I worry the process might get terminated by the crash before 50 resizes, maybe lower this a bit?

Reply:
I based this on some of the crash logs we're seeing - `updateVisibleCells` seems to get called 100 times before UICV gives up and crashes. I was thinking half of Apple's layout recursion detection threshold would work well and also filter out anything that could happen due to device rotation / other legitimate resizes.

For example:

```
NSInternalInconsistencyException: UICollectionView (<EpoxyCollectionView.CollectionView 0x11089ec00>) is stuck in its update/layout loop. This can happen for many reasons, including self-sizing views whose preferred attributes are not returning a consistent size. To debug this issue, check the Console app for logs in the "UICollectionViewRecursion" category.

0   CoreFoundation      ___exceptionPreprocess
1   libobjc.A.dylib     _objc_exception_throw
2   Foundation          -[NSAssertionHandler handleFailureInMethod:object:file:lineNumber:description:]
3   UIKitCore           -[UICollectionView _setNeedsVisibleCellsUpdate:withLayoutAttributes:]
4   UIKitCore           -[UICollectionView _invalidateLayoutWithContext:]
5   UIKitCore           -[UICollectionViewLayout invalidateLayoutWithContext:]
6   Airbnb              invalidateLayout (MagazineLayout.swift:880:11)
7   Airbnb              invalidateLayout (<compiler-generated>)
8   UIKitCore           -[UICollectionView _checkForPreferredAttributesInView:originalAttributes:]
9   UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
10  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
11  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
12  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
13  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
14  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
15  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
16  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
17  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
18  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
19  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
20  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
21  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
22  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
23  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
24  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
25  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
26  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
27  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
28  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
29  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
30  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
31  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
32  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
33  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
34  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
35  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
36  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
37  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
38  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
39  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
40  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
41  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
42  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
43  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
44  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
45  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
46  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
47  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
48  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
49  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
50  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
51  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
52  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
53  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
54  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
55  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
56  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
57  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
58  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
59  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
60  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
61  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
62  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
63  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
64  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
65  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
66  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
67  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
68  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
69  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
70  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
71  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
72  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
73  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
74  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
75  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
76  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
77  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
78  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
79  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
80  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
81  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
82  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
83  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
84  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
85  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
86  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
87  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
88  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
89  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
90  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
91  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
92  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
93  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
94  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
95  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
96  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
97  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
98  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
99  UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
100 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
101 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
102 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
103 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
104 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
105 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
106 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
107 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
108 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
109 UIKitCore           -[UICollectionView _updateVisibleCellsNow:]
110 UIKitCore           -[UICollectionView layoutSubviews]
111 UIKitCore           -[UIView(CALayerDelegate) layoutSublayersOfLayer:]
112 QuartzCore          CA::Layer::layout_if_needed(CA::Transaction*)
113 QuartzCore          CA::Layer::layout_and_display_if_needed(CA::Transaction*)
114 QuartzCore          CA::Context::commit_transaction(CA::Transaction*, double, double*)
115 QuartzCore          CA::Transaction::commit()
116 QuartzCore          CA::Transaction::flush_as_runloop_observer(bool)
117 CoreFoundation      ___CFRUNLOOP_IS_CALLING_OUT_TO_AN_OBSERVER_CALLBACK_FUNCTION__
118 CoreFoundation      ___CFRunLoopDoObservers
119 CoreFoundation      ___CFRunLoopRun
120 CoreFoundation      _CFRunLoopRunSpecific
121 GraphicsServices    _GSEventRunModal
122 UIKitCore           -[UIApplication _run]
123 UIKitCore           _UIApplicationMain
124 libswiftUIKit.dylib UIApplicationMain(_:_:_:_:)
125 Airbnb              main (<compiler-generated>)
126 Airbnb              $main (AppDelegate.swift:18:1)
127 dyld                start
```

Reply:
Played around with it some more, and it seems like if I lower the # of sizing attempts from 50s to something like 20s, then I can also lower the timing threshold from 2s down to something like 0.5s.

Not sure what the right balance is here. I was thinking it might be safer to pick higher numbers that way we don't assert in a situation that wouldn't actually cause the UICV crash. Thoughts?

Reply:
@erichoracek @thedrick thoughts on striking the right balance here? The threshold of 50 and 2s seem to work well, but I could reduce both numbers if you think that's better.

Reply:
If you feel comfortable with 2 second and 50 attempts I'm ok with it as well 👍  I appreciate you testing this to come up with a reasonable number